zerver/lib/thumbnail.py

```diff
@@ -21,6 +26,83 @@
 # Reject uploaded emoji which are larger than this
 MAX_EMOJI_GIF_FILE_SIZE_BYTES = 128 * 1024 * 1024  # 128 kb
 
+# These are the image content-types which the server supports parsing
+# and thumbnailing; these do not need to supported on all browsers,
+# since we will the serving thumbnailed versions of them.
+THUMBNAIL_ACCEPT_IMAGE_TYPES = frozenset(
+    [
+        "image/avif",
+        "image/gif",
+        "image/heic",
+        "image/jpeg",
+        "image/png",
+        "image/svg+xml",
+        "image/webp",
+    ]
+)
+
+T = TypeVar("T", bound="BaseThumbnailFormat")
+
+
+@dataclass(frozen=True, eq=True)
+class BaseThumbnailFormat:
+    extension: str
+    max_width: int
+    max_height: int
+    animated: bool
+
+    @override
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, BaseThumbnailFormat):
+            return False
+        return str(self) == str(other)
+
+    @override
+    def __str__(self) -> str:
+        animated = "-anim" if self.animated else ""
+        return f"{self.max_width}x{self.max_height}{animated}.{self.extension}"
+
+    @classmethod
+    def from_string(cls: Type[T], format_string: str) -> Optional[T]:
+        format_parts = re.match(r"(\d+)x(\d+)(-anim)?\.(\w+)$", format_string)
+        if format_parts is None:
+            return None
+
+        return cls(
+            max_width=int(format_parts[1]),
+            max_height=int(format_parts[2]),
+            animated=format_parts[3] is not None,
+            extension=format_parts[4],
+        )
+
+
+@dataclass(frozen=True, eq=True)
+class ThumbnailFormat(BaseThumbnailFormat):
+    opts: Optional[str] = ""
+
+
+# Note that this is serialized into a JSONB column in the database,
+# and as such fields cannot be removed without a migration.
+@dataclass(frozen=True, eq=True)
+class StoredThumbnailFormat(BaseThumbnailFormat):
+    content_type: str
+    width: int
+    height: int
+    byte_size: int
+
+
+# Formats that we generate; the first animated and non-animated
+# options on this list are the ones which are written into
+# rendered_content.
+THUMBNAIL_OUTPUT_FORMATS = [
+    ThumbnailFormat("webp", 150, 100, animated=True),
+    ThumbnailFormat("webp", 150, 100, animated=False),
+    ThumbnailFormat("avif", 150, 100, animated=False),
+    ThumbnailFormat("jpg", 150, 100, animated=False),
+    ThumbnailFormat("avif", 600, 400, animated=False),
+    ThumbnailFormat("webp", 600, 400, animated=False),
+]
```

Comment:
Similarly, we should probably comment-document why `avif/webp` have larger pixel sizes.

Reply:
Honestly, these were just tossed in there to test the differing sizes.  We're not currently using these 600x400 images anywhere, and we can potentially just drop them until something (e.g. mobile) wants to use them.